processing/src/main/java/org/apache/druid/frame/write/cast/TypeCastSelectors.java

```diff
@@ -0,0 +1,149 @@
+   *
+   * @param columnSelectorFactory underlying factory
+   * @param column                column name
+   * @param desiredType           desired type of selector. Can be anything except {@link ColumnType#STRING}.
+   *                              For strings, use {@link DimensionSelector} rather than {@link ColumnValueSelector}.
+   */
+  public static ColumnValueSelector<?> makeColumnValueSelector(
+      final ColumnSelectorFactory columnSelectorFactory,
+      final String column,
+      final ColumnType desiredType
+  )
+  {
+    final ColumnValueSelector<?> selector = columnSelectorFactory.makeColumnValueSelector(column);
+    final ColumnCapabilities selectorCapabilities = columnSelectorFactory.getColumnCapabilities(column);
+
+    return wrapColumnValueSelectorIfNeeded(
+        selector,
+        selectorCapabilities,
+        columnSelectorFactory.getRowIdSupplier(),
+        desiredType
+    );
+  }
+
+  public static ColumnValueSelector<?> wrapColumnValueSelectorIfNeeded(
+      final ColumnValueSelector<?> selector,
+      @Nullable final ColumnCapabilities selectorCapabilities,
+      @Nullable final RowIdSupplier rowIdSupplier,
+      final ColumnType desiredType
+  )
+  {
+    if (desiredType.is(ValueType.STRING)) {
+      throw DruidException.defensive("Type[%s] should be read using a DimensionSelector", desiredType);
+    } else if (desiredType.isNumeric()
+               && (selectorCapabilities == null || !selectorCapabilities.isNumeric())) {
+      // When capabilities are unknown, or known to be non-numeric, fall back to getObject() and explicit typecasting.
+      // This avoids using primitive numeric accessors (getLong / getDouble / getFloat / isNull) on a selector that
+      // may not support them.
+      return new ObjectToNumberColumnValueSelector(selector, desiredType, rowIdSupplier);
+    } else if (desiredType.isArray()) {
+      // Always wrap if desiredType is an array. Even if the underlying selector claims to offer the same type as
+      // desiredType, it may fail to respect the BaseObjectColumnValueSelector contract. For example, it may return
+      // List rather than Object[]. (RowBasedColumnSelectorFactory can do this if used incorrectly, i.e., if the
+      // ColumnInspector declares type ARRAY<X> for a column, but the RowAdapter does not provide Object[].)
+      return new ObjectToArrayColumnValueSelector(selector, desiredType, rowIdSupplier);
+    } else {
+      // OK to return the original selector.
+      return selector;
+    }
+  }
+
+  /**
+   * Coerce an object to an object compatible with what {@link BaseObjectColumnValueSelector#getObject()} for a column
+   * of the provided desiredType. Never throws an exception. If coercion fails, replaces the object that failed to
+   * coerce with null.
+   *
+   * @param obj         object
+   * @param desiredType desired type
+   */
+  @Nullable
+  public static Object bestEffortCoerce(
```

Comment:
this method seems odd to me, it is basicaly like `ExprEval.ofType` which separates handling based on the expected `ExpressionType` (instead of `ColumnType` like is used here), but then uses `ExprEval.bestEffortOf` inside of the cases (which uses a much larger set of instanceof checks than `ExprEval.ofType`).

What sets this method apart from the others?

Reply:
I pushed a new commit that replaces this method with `ExprEval.ofType`. It seems to do what we need. I didn't try it at first, because I misunderstood what it does— I didn't realize it also included coercion. I added javadoc to `ExprEval.ofType` explaining that it does do coercion.